src/data/hostUrlSingleton.ts

```diff
@@ -0,0 +1,89 @@
+import type { ConfigurationParameters as IndexOperationsApiConfigurationParameters } from '../pinecone-generated-ts-fetch';
+import {
+  IndexOperationsApi,
+  Configuration as ApiConfiguration,
+} from '../pinecone-generated-ts-fetch';
+import type { PineconeConfiguration } from './types';
+import { middleware } from '../utils/middleware';
+import { queryParamsStringify, buildUserAgent, getFetch } from '../utils';
+import { describeIndex } from '../control';
+import { PineconeUnableToResolveHostError } from '../errors';
+
+// We use describeIndex to retrieve the data plane url (host) for a given API key
+// and index. We only ever want to call describeIndex a maximum of once per API key
+// and index so we cache them in a singleton for reuse.
+export const HostUrlSingleton = (function () {
+  const hostUrls = {}; // map of apiKey-indexName to hostUrl
+  let indexOperationsApi: InstanceType<typeof IndexOperationsApi> | null = null;
+
+  const _buildIndexOperationsApi = (config: PineconeConfiguration) => {
+    const { apiKey } = config;
+    const controllerPath = `https://api.pinecone.io`;
+    const apiConfig: IndexOperationsApiConfigurationParameters = {
+      basePath: controllerPath,
+      apiKey,
+      queryParamsStringify,
+      headers: {
+        'User-Agent': buildUserAgent(false),
+      },
+      fetchApi: getFetch(config),
+      middleware,
+    };
+
+    return new IndexOperationsApi(new ApiConfiguration(apiConfig));
+  };
+
+  const _describeIndex = async (
+    config: PineconeConfiguration,
+    indexName: string
+  ): Promise<string> => {
+    if (!indexOperationsApi) {
+      indexOperationsApi = _buildIndexOperationsApi(config);
+    }
+
+    const describeResponse = await describeIndex(indexOperationsApi)(indexName);
+    const host = describeResponse.status?.host;
+
+    if (!host) {
```

Comment:
It's fine to cover this case here because the type indicates it could occur (mostly because of loose typing in the openapi spec, I think), but in practice I have never seen the host value come back missing. I think the url is known immediately, whether or not the index is actually ready to receive data plane calls.

Reply:
Yup, primarily we're satisfying flow. Maybe we could opt to update the OpenAPI spec to make this a bit easier for us.